doc/mgmt/config_replace_rollback_HLD.md

```diff
@@ -0,0 +1,416 @@
+- "commit confirm" command to confirm new configuration changes that are applied to running-configuration, and move from trial to permanent stage. Chekcpoints will be generated by configure session. 
+
+  
+### 1.3.2 Container
+No new containers are introduced for this feature. Existing mgmt-framework container is updated.
+
+### 1.3.3 SAI Overview
+No new or existing SAI services are required
+
+
+# 2 Functionality
+
+Following functionalities are provided to support Config Replace and Rollback.
+- timer goroutine started by a timer option
+- automatic revert to the saved backup configuration if a timer gets expired
+- revert back to previous configuration state if "kill config session" is executed
+- cancel a timer 
+- reject the operation if candidate datastore is dirtified
+- action to support both session and session-less
+- action to archive CONFIG_DB to the backup configuration for rollback
+- modeless rpc "commit" to support timer
+- modeless rpc "commit" to to support confirm
+- add a rule to rbac to allow only admin user
+- add support to retrieve configuration file from remote server and decompress the file if it is in compressed format
+- error reporting
+- debug utilities
+
+# 3 Design
+## 3.1 Overview
+A new config replacement componenent on the server will be a key implemenation to orchestrate config replace operation requested by different NBIs. It loads a desired config file specified by a request and compares against CONFIG_DB to generate the diff results tagged with operation, table, keys and fields. The resultant diff data set will be sorted by operation - DELETE, CREATE and REPLACE in this order - as we do for REST/gNMI operation, and subsequently sorted by table orders learnt from CVL to apply the changes as a single transaction to CVL/DB layer.
+
+Currently the ordering for target tables transformed for REST/gNMI path uses CVL api cvlSess.GetOrderedDepTables() to learn table dependency within a given module. But, the diff'ed dataset from a serialized json uses cvlSess.GetDepTables() to learn table dependencies across SONiC yang modules to sort all the tables from diffy operation. 
+
+### 3.1.1 Exec mode
+When "copy <source-url> running-configuration replace" is executed from exec mode, CLI sends a REST POST method on "copy" RPC to the server. Unlike configure session mode, the session token is not assigned to the request and the configuration changes will not be cached, but directly committed to CONFIG_DB after going through diffy, sorting, validation process. 
+
+If a configure session is already active, the command will be rejected with a warning "Configure session is in progress." 
+
+If CONFIG_DB is being locked by other configuration commands, it will be also rejected. 
+
+### 3.1.2 Configure session mode
```

Comment:
What is the behavior if the CLI session got disconnected while the commit replace is happening? After the user re-login, will the user be able to tell whether his previous commit replace goes through or not? 

In the case where there is an error to the new config the user tries to apply, will the user get the same error reporting when commit replace is executed in exec mode vs. config session mode?

Reply:
I think your question is about a vty session disconnected during operation.
Indeed there are multiple scenarios.
- if a vty session gets disconnected during "replace", all the changes are still in candidate datastore like any other changes
- if a vty session gets disconnected right after a user issues commit **without** a timer, a re-login user can find a cli log to see the commit succeeds or not. Assuming no chance to see a success or failure before it gets disconnected.
- if CLI session gets disconnected right after a user issues commit **with** a timer, a re-login user can find a cli log to see the commit succeeds or not. If commit errored, a user can abort. If it succeeds, a user can reenter the session to commit before a timer gets expired. If a timer already gets expired, the CONFIG DB is auto-reverted to the previous stage.
- In exec mode, it is single-shot action. A user can see a cli-log.